### PR TITLE
[core] User can specify variables by ID in Model::reshape as optional paramter

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - The `Model::reshape` function has been extended by additional parameter which allow user specify new variables shapes by ID.
 - Specify only inputs shape can be not enough in model which use Variables. Variable specify range of shape inferred form initial model. New input shape can be outside this range and then user has to specify  new Variable shape.
 - Provide only shapes for Parameters can be not enough after MakeStateful transformation and Variables shape has to be update also.
 - The Kaldi e2e test can  be updated to use new API to slave test issues.

### Tickets:
 - Can fix [CVS-126178](https://jira.devtools.intel.com/browse/CVS-126178)

### CODE REVIEW

File: src/core/src/op/read_value.cpp [hunk 1]
```diff
@@ -134,6 +134,10 @@ bool ReadValue::visit_attributes(AttributeVisitor& visitor) {
 }
 
 void ReadValue::revalidate_and_infer_types() {
+    // Update variable shape if not extend new shape (e.g reshape result).
+    if (get_input_size() > 0 && !m_variable->get_info().data_shape.relaxes(get_input_partial_shape(0))) {
+        m_variable->update_data_shape(get_input_partial_shape(0));
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/core/src/op/read_value.cpp:139]: according to the previous changes, it's not allowed to update Variable shape.
Variable shape is like a mask for Input shape.
the corresponding dim have to be in range that Variable dim defines. 
If we would like to set batch = 2 for the input, the corresponding batch dim of Variable shape have to be dynamic, e.g 1:2

[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/op/read_value.cpp:139]: The failure is from e2e tests. At start the model is create with batch dim 1 but then its reshaped to have batch 2.
How it should be solved to run these test correctly?

**Comment by slyalin** [comment]
@praasz, please comment on value outside Kaldi. Kaldi support was stopped a year ago. I see setting shapes for variables reasonable in the context that we started to produce stateful LLMs, but we had a way to set proper shapes for variables with ReadValue initializer.

**Comment by itikhono** [comment]
> @praasz, please comment on value outside Kaldi. Kaldi support was stopped a year ago. I see setting shapes for variables reasonable in the context that we started to produce stateful LLMs, but we had a way to set proper shapes for variables with ReadValue initializer.

Not only Kaldi models may become non-reshapable if they use Variables.
 
E.g to reshape model after applying MakeStateful transformation, usually it's not enough to set new shapes for Parameters, rather, users need to update Variable shapes as well. MakeStateful is a common transformation, not only for Kaldi
 
Without these changes in "reshape" method, users need to search and update Variables manually.

About "ReadValue initializer", MakeStateful transformation doesn't create this initializer. As far as I remember, these initializers for LLMs were manually created.

**Review by jiwaszki — COMMENTED** [review]
LGTM on bindings side in current form.

**Comment by mlukasze** [comment]
are you ok with merge it, or you have specific change requests that should wait for Pawel fixes?

**Comment by praasz** [comment]
@itikhono , @jane-intel could you review the PR?

File: src/core/src/model.cpp [hunk 2]
```diff
@@ -818,17 +855,43 @@ void ov::Model::reshape(const std::map<ov::Output<ov::Node>, ov::PartialShape>&
     if (!need_reshape)
         return;
 
+    // const auto& variables = get_variables();
```
Comment (jane-intel) [inline src/core/src/model.cpp:858]: ?

**Review by jane-intel — COMMENTED** [review]
General questions:

- Why did we pick to use strings to identify variables? We could've used indices or instances instead.

- Why do we add 4 methods instead of adding optional parameters to existing ones?

File: src/core/src/model.cpp (hunk 2 above)
Comment (praasz) [reply src/core/src/model.cpp:858]: Will remove it.

**Comment by praasz** [comment]
> General questions:
> 
> * Why did we pick to use strings to identify variables? We could've used indices or instances instead.
It looks like name is user friendly to identify it.

The indices could be used but I think is more enigmatic for user and I'm not sure if will easy identify it in model if required variable will be always at same position.
The instances is like some kind of ID (address) which value is even more not friendly. The get variables by instance there is function `get_variables()`. Get and use instance will allow to modify shapes before input reshape which can result with error in validation:
```cpp
        ov::pass::Manager ssr_manager;
        ssr_manager.register_pass<ov::pass::SmartReshape>();
        // Here the old shapes for inputs and variables has to be used, if variables reshaped then could be error here on validation.
        ssr_manager.run_passes(shared_from_this());

        // Here new shapes for inputs and variables can be applied and validated.
        reshape_only(new_param_shapes, new_vars_shapes);
```

But If you got ideas how used no string ids but indices or instances which will be useful I can adjust implementation. I would recommend to limit the `ID` type to minimum to not produce to many reshape overloads.

> * Why do we add 4 methods instead of adding optional parameters to existing ones?

Is done to not create the default map on each call (is internally created only once and reference used instead).
If required it can be changed to use default parameter instead.
> 
> Do we need an architecture review? We are changing the public API here, it is required.

Will check and can  wait with merge when changes will be approved.

**Review by ilya-lavrenov — APPROVED** [review]

File: src/core/include/openvino/core/model.hpp [hunk 3]
```diff
@@ -140,10 +140,14 @@ class OPENVINO_API Model : public std::enable_shared_from_this<Model> {
...
     ov::Output<ov::Node> add_output(const ov::Output<ov::Node>& port);
 
-    void reshape(const ov::PartialShape& partial_shape);
-    void reshape(const std::map<size_t, ov::PartialShape>& partial_shapes);
-    void reshape(const std::map<std::string, ov::PartialShape>& partial_shapes);
-    void reshape(const std::map<ov::Output<ov::Node>, ov::PartialShape>& partial_shapes);
+    void reshape(const ov::PartialShape& partial_shape,
+                 const std::unordered_map<std::string, ov::PartialShape>& variable_shapes = {});
+    void reshape(const std::map<size_t, ov::PartialShape>& partial_shapes,
+                 const std::unordered_map<std::string, ov::PartialShape>& variable_shapes = {});
```
Comment (itikhono) [inline src/core/include/openvino/core/model.hpp:146]: minor: should it be `map` instead of `unordered_map` to be aligned with 1st argument?

Comment (praasz) [reply src/core/include/openvino/core/model.hpp:146]: I think is not required to have same map type. 
as for inputs shapes the map will be sorted by port number which give some intuition which port will get new shape.
The key for variables is name and if sorted I think is not exactly how they are stored internally. The unordered map will be simpler (faster) for creation.

**Pull request merged**